subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyExcludeResolveIntegrationTest.groovy

```diff
@@ -91,4 +92,52 @@ task check {
         'non-matching module attribute'                    | [module: 'unknown']                       | ['test-1.45.jar', 'foo-2.0.jar', 'bar-3.0.jar', 'company-4.0.jar', 'other-company-4.0.jar', 'enterprise-5.0.jar', 'baz-6.0.jar']
         'attempting to exclude declared module'            | [group: 'org.gradle', module: 'test']     | ['test-1.45.jar', 'foo-2.0.jar', 'bar-3.0.jar', 'company-4.0.jar', 'other-company-4.0.jar', 'enterprise-5.0.jar', 'baz-6.0.jar']
     }
+
+    /**
+     * Dependency graph:
+     *
+     * org.gradle:test:1.0
+     * +--- org.foo:foo:2.0
+     *      \--- org.bar:bar:3.0
+     */
+    @Unroll
+    @Issue("gradle/gradle#951")
+    def "fine grain transitive dependency #condition"() {
+        given:
+        def testModule = mavenRepo().module('org.gradle', 'test', '1.0')
+        def fooModule = mavenRepo().module('org.foo', 'foo', '2.0')
+        def barModule = mavenRepo().module('org.bar', 'bar', '3.0')
+
+        barModule.publish()
```

Comment:
Let's remove the new lines between the `publish()` statements and the Spock `and` statement. It might be more readable if we just roll it into one `given`.